src/core_tests/determinism.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    pollers::MockServerGatewayApis,
+    prototype_rust_sdk::{TestRustWorker, WfContext, WorkflowResult},
+    test_help::{
+        build_mock_pollers, canned_histories, mock_core, MockPollCfg, ResponseType,
+        DEFAULT_WORKFLOW_TYPE, TEST_Q,
+    },
+};
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+use temporal_sdk_core_protos::temporal::api::enums::v1::WorkflowTaskFailedCause;
+
+static DID_FAIL: AtomicBool = AtomicBool::new(false);
+pub async fn timer_wf_fails_once(mut ctx: WfContext) -> WorkflowResult<()> {
```

Comment:
Why is it important to only fail once? One might expect a deterministic workflow to always panic given the same inputs, so the test might as well too, right?

Reply:
Because we're testing nondeterministic behavior here. We want to see that a workflow can fail a WFT for spurious reasons, send the right type of failure, and then recover the next go-round.